### ISSUE TITLE
gopy: concurrent map writes during tests

### ISSUE THREAD

**Issue by sbinet** [go-python/gopy#159]
Commit ab145c0 that introduced support for test coverage, also introduced a regression:

```
fatal error: concurrent map writes
goroutine 12 [running]:
runtime.throw(0x715aeb, 0x15)
	/home/travis/.gimme/versions/go1.7.6.linux.amd64/src/runtime/panic.go:566 +0x95 fp=0xc420140570 sp=0xc420140550
runtime.mapassign1(0x6c4160, 0xc420146d50, 0xc420140748, 0xc420140710)
	/home/travis/.gimme/versions/go1.7.6.linux.amd64/src/runtime/hashmap.go:553 +0x2e1 fp=0xc420140658 sp=0xc420140570
github.com/gonuts/flag.(*FlagSet).parseOne(0xc42000edc0, 0x6da101, 0x0, 0x0)
	/home/travis/gopath/src/github.com/gonuts/flag/flag.go:749 +0x345 fp=0xc420140868 sp=0xc420140658
github.com/gonuts/flag.(*FlagSet).Parse(0xc42000edc0, 0xc4200a96d0, 0x3, 0x3, 0x85d980, 0xc4200b1d40)
	/home/travis/gopath/src/github.com/gonuts/flag/flag.go:761 +0x60 fp=0xc420140898 sp=0xc420140868
github.com/gonuts/commander.(*Command).Dispatch(0xc42000ed80, 0xc4200a96d0, 0x3, 0x3, 0x1, 0x0)
	/home/travis/gopath/src/github.com/gonuts/commander/commands.go:215 +0x2ec fp=0xc420140930 sp=0xc420140898
github.com/gonuts/commander.(*Command).Dispatch(0xc42000eea0, 0xc4200a96c0, 0x4, 0x4, 0x0, 0x0)
	/home/travis/gopath/src/github.com/gonuts/commander/commands.go:187 +0x6cf fp=0xc4201409c8 sp=0xc420140930
github.com/go-python/gopy.run(0xc4200a96c0, 0x4, 0x4, 0x7120b1, 0xc)
	/home/travis/gopath/src/github.com/go-python/gopy/main.go:42 +0x9e fp=0xc420140a30 sp=0xc4201409c8
github.com/go-python/gopy.testPkgBackend(0xc420074540, 0x70ef62, 0x3, 0x70ffaf, 0x7, 0x7120b1, 0xc, 0xc420010b00, 0x1, 0x1, ...)
	/home/travis/gopath/src/github.com/go-python/gopy/main_test.go:127 +0x2c5 fp=0xc420140ef8 sp=0xc420140a30
github.com/go-python/gopy.testPkg.func1(0xc420074540)
	/home/travis/gopath/src/github.com/go-python/gopy/zsupport_test.go:28 +0x95 fp=0xc420140f78 sp=0xc420140ef8
testing.tRunner(0xc420074540, 0xc420025c20)
	/home/travis/.gimme/versions/go1.7.6.linux.amd64/src/testing/testing.go:610 +0x81 fp=0xc420140fa0 sp=0xc420140f78
runtime.goexit()
	/home/travis/.gimme/versions/go1.7.6.linux.amd64/src/runtime/asm_amd64.s:2086 +0x1 fp=0xc420140fa8 sp=0xc420140fa0
created by testing.(*T).Run
	/home/travis/.gimme/versions/go1.7.6.linux.amd64/src/testing/testing.go:646 +0x2ec

```

**Comment by sbinet** [comment]
@rameshg87 do you want to take a stab at this?

**Comment by initialed85** [comment]
Alright, I think I'm getting sane results now- Go1.10 was really causing me some problems with the way it caches test results; after reverting to Go1.9, I was able to my tests failing reliably with https://github.com/go-python/gopy/commit/c99fa35fcb039b8b3b70f1c7481a06241dc0375f.

Ultimately, the issue looks to be related to the app entrypoint (`run()` in `main.go`) expecting to be the only one of itself running; running the tests with `-parallel=1` stops the issue happening.

I found two places that I was able to adjust to ensure tests pass with `parallel=8`

- `main.go`
    - change the `app` variable of type `&commander.Command` to be local to `run()`
- `package.go`
    - protect the assignment of `universe.pkg` in `NewPackage()` with a mutex

The solution is maybe not 100% perfect (i.e. we might stumble across goroutine safety issues with tests that are yet to be written) but I figure the application is only tested like this, not operated like this.

But I'm going to check the blazes out of it to make sure I'm not celebrating preemptively again!

If all is well- should my fixed PR include the commits that were reverted? Or will we bring them in later?
